File: services/text_parser.py
"""智能文本解析服务"""

import re
from typing import Dict, List, Optional
from utils.validators import (
    extract_url,
    extract_api_key,
    is_valid_url,
    is_valid_api_key,
)
from utils.vendor_detector import detect_vendor


class TextParser:
    """智能文本解析器"""

    def __init__(self):
        pass

    def parse(self, text: str) -> Dict:
        """
        解析文本，提取URL和API Key

        Args:
            text: 输入文本

        Returns:
            {
                'base_url': 'https://...',  # 最佳匹配
                'api_key': 'sk-...',        # 最佳匹配
                'vendor': 'openai',
                'url_candidates': [...],     # 所有候选URL
                'key_candidates': [...]      # 所有候选Key
            }
        """
        result = {
            "base_url": None,
            "api_key": None,
            "vendor": "custom",
            "url_candidates": [],
            "key_candidates": [],
        }

        # 1. 查找所有可能的URL
        url_candidates = self._find_url_candidates(text)
        result["url_candidates"] = url_candidates

        # 选择最佳URL
        if url_candidates:
            result["base_url"] = url_candidates[0]["value"]

        # 2. 查找所有可能的API Key
        key_candidates = self._find_key_candidates(text)
        result["key_candidates"] = key_candidates

        # 选择最佳Key
        if key_candidates:
            result["api_key"] = key_candidates[0]["value"]

        # 3. 推断厂商
        if result["base_url"]:
            result["vendor"] = detect_vendor(result["base_url"])

        return result

    def _find_url_candidates(self, text: str) -> List[Dict]:
        """
        查找所有可能的URL

        Returns:
            [{'value': 'https://...', 'score': 0.9}, ...]
        """
        candidates = []

        # 匹配所有URL
        pattern = r'https?://[^\s,;。，；\'"]+(?:/[^\s,;。，；\'"]*)?'
        matches = re.finditer(pattern, text)

        for match in matches:
            url = match.group(0)
            # 清理末尾标点
            url = url.rstrip(".,;:!?。，；：！？")

            if is_valid_url(url):
                score = self._score_url(url)
                candidates.append({"value": url, "score": score})

        # 按分数排序
        candidates.sort(key=lambda x: x["score"], reverse=True)

        # 去重
        seen = set()
        unique_candidates = []
        for c in candidates:
            if c["value"] not in seen:
                seen.add(c["value"])
                unique_candidates.append(c)

        return unique_candidates

    def _find_key_candidates(self, text: str) -> List[Dict]:
        """
        查找所有可能的API Key

        Returns:
            [{'value': 'sk-...', 'score': 0.9}, ...]
        """
        candidates = []

        # 1. 查找sk-开头的（OpenAI格式）
        sk_pattern = r"sk-[A-Za-z0-9]{20,}"
        for match in re.finditer(sk_pattern, text):
            key = match.group(0)
            candidates.append(
                {
                    "value": key,
                    "score": 0.95,  # 高置信度
                }
            )

        # 2. 查找Bearer Token
        bearer_pattern = r"Bearer\s+([A-Za-z0-9_-]{20,})"
        for match in re.finditer(bearer_pattern, text, re.IGNORECASE):
            key = match.group(1)
            candidates.append({"value": key, "score": 0.9})

        # 3. 查找其他长字符串（排除URL）
        lines = text.split("\n")
        for line in lines:
            line = line.strip()
            # 跳过URL行
            if line.startswith("http"):
                continue
            # 跳过包含中文的行
            if re.search(r"[\u4e00-\u9fff]", line):
                continue
            # 匹配32-100个字符的字母数字组合
            if 32 <= len(line) <= 100 and re.match(r"^[A-Za-z0-9_-]+$", line):
                candidates.append(
                    {
                        "value": line,
                        "score": 0.7,  # 中等置信度
                    }
                )

        # 按分数排序
        candidates.sort(key=lambda x: x["score"], reverse=True)

        # 去重
        seen = set()
        unique_candidates = []
        for c in candidates:
            if c["value"] not in seen:
                seen.add(c["value"])
                unique_candidates.append(c)

        return unique_candidates

    def _score_url(self, url: str) -> float:
        """
        给URL打分

        评分规则：
        - 包含/v1: +0.3
        - HTTPS: +0.1
        - 已知厂商域名: +0.2
        - 基础分: 0.4
        """
        score = 0.4

        url_lower = url.lower()

        # 包含API版本号
        if "/v1" in url_lower:
            score += 0.3

        # HTTPS
        if url_lower.startswith("https://"):
            score += 0.1

        # 已知厂商
        known_domains = [
            "openai.com",
            "anthropic.com",
            "googleapis.com",
            "deepseek",
            "zhipuai",
            "moonshot",
        ]
        if any(domain in url_lower for domain in known_domains):
            score += 0.2

        return min(score, 1.0)

File: src-tauri/src/services/text_parser.rs
use crate::utils::vendor_detector::detect_vendor;
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Candidate {
    pub value: String,
    pub score: f32,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ParseResult {
    pub base_url: Option<String>,
    pub api_key: Option<String>,
    pub vendor: String,
    pub url_candidates: Vec<Candidate>,
    pub key_candidates: Vec<Candidate>,
}

pub struct TextParser;

impl TextParser {
    pub fn parse(text: &str) -> ParseResult {
        let url_candidates = Self::find_url_candidates(text);
        let key_candidates = Self::find_key_candidates(text);

        let base_url = url_candidates.first().map(|c| c.value.clone());
        let api_key = key_candidates.first().map(|c| c.value.clone());

        let vendor = if let Some(ref url) = base_url {
            detect_vendor(url)
        } else {
            "custom".to_string()
        };

        ParseResult {
            base_url,
            api_key,
            vendor,
            url_candidates,
            key_candidates,
        }
    }

    fn find_url_candidates(text: &str) -> Vec<Candidate> {
        let re = Regex::new(r#"https?://[^\s,;。，；'"\]]+(?:/[^\s,;。，；'"\]]*)?"#).unwrap();
        let mut candidates = Vec::new();

        for mat in re.find_iter(text) {
            let mut url = mat.as_str().to_string();
            // Trim trailing punctuation
            url = url
                .trim_end_matches(|c: char| ".,;:!?。，；：！？".contains(c))
                .to_string();

            if !url.is_empty() {
                let score = Self::score_url(&url);
                candidates.push(Candidate { value: url, score });
            }
        }

        candidates.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap());
        // Deduplicate
        let mut seen = std::collections::HashSet::new();
        candidates.retain(|c| seen.insert(c.value.clone()));

        candidates
    }

    fn find_key_candidates(text: &str) -> Vec<Candidate> {
        let mut candidates = Vec::new();

        // 1. OpenAI format sk-
        let sk_re = Regex::new(r"sk-[A-Za-z0-9]{20,}").unwrap();
        for mat in sk_re.find_iter(text) {
            candidates.push(Candidate {
                value: mat.as_str().to_string(),
                score: 0.95,
            });
        }

        // 2. Bearer Token
        let bearer_re = Regex::new(r"(?i)Bearer\s+([A-Za-z0-9_-]{20,})").unwrap();
        for cap in bearer_re.captures_iter(text) {
            if let Some(mat) = cap.get(1) {
                candidates.push(Candidate {
                    value: mat.as_str().to_string(),
                    score: 0.9,
                });
            }
        }

        // 3. Other long strings
        for line in text.lines() {
            let line = line.trim();
            if line.starts_with("http") {
                continue;
            }
            if line.chars().any(|c| ('\u{4e00}'..='\u{9fff}').contains(&c)) {
                continue;
            }

            if line.len() >= 32 && line.len() <= 100 {
                let long_str_re = Regex::new(r"^[A-Za-z0-9_-]+$").unwrap();
                if long_str_re.is_match(line) {
                    candidates.push(Candidate {
                        value: line.to_string(),
                        score: 0.7,
                    });
                }
            }
        }

        candidates.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap());
        let mut seen = std::collections::HashSet::new();
        candidates.retain(|c| seen.insert(c.value.clone()));

        candidates
    }

    pub fn parse_model_names(text: &str) -> Vec<String> {
        let patterns = [
            r"(?i)gpt-[\w.-]+",
            r"(?i)claude-[\w.-]+",
            r"(?i)gemini-[\w.-]+",
            r"(?i)deepseek-[\w.-]+",
            r"(?i)glm-[\w.-]+",
            r"(?i)moonshot-[\w.-]+",
            r"[\w-]+-\d{8,}",
            r"(?i)o1-[\w.-]+",
            r"(?i)text-[\w.-]+",
        ];

        let mut models = std::collections::HashSet::new();

        for pattern in patterns {
            let re = Regex::new(pattern).unwrap();
            for mat in re.find_iter(text) {
                let model = mat.as_str().to_lowercase();
                if model.len() >= 3 && model.len() <= 80 {
                    models.insert(model);
                }
            }
        }

        let mut result: Vec<String> = models.into_iter().collect();
        result.sort();
        result
    }

    fn score_url(url: &str) -> f32 {
        let mut score: f32 = 0.4;
        let url_lower = url.to_lowercase();

        if url_lower.contains("/v1") {
            score += 0.3;
        }

        if url_lower.starts_with("https://") {
            score += 0.1;
        }

        let known_domains = [
            "openai.com",
            "anthropic.com",
            "googleapis.com",
            "deepseek",
            "zhipuai",
            "moonshot",
        ];
        if known_domains.iter().any(|&d| url_lower.contains(d)) {
            score += 0.2;
        }

        score.min(1.0)
    }
}
